src/runtime_src/driver/xclng/drm/xocl/subdev/nifd.c

```diff
@@ -0,0 +1,717 @@
+    .owner = THIS_MODULE,
+    .open = char_open,
+    .release = char_close,
+    .unlocked_ioctl = nifd_ioctl,
+};
+
+static int nifd_probe(struct platform_device *pdev)
+{
+    struct xocl_nifd *nifd;
+    struct resource *res;
+    struct xocl_dev_core *core;
+    struct FeatureRomHeader rom;
+    bool nifd_valid = false;
+    int err = 0;
+
+    nifd = xocl_drvinst_alloc(&pdev->dev, sizeof(*nifd));
+    if (!nifd)
+        return -ENOMEM;
+
+    res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+    nifd->nifd_base = ioremap_nocache(res->start, 
+                        res->end - res->start + 1);
+    if (!nifd->nifd_base)
+    {
+        err = -EIO;
+        xocl_err(&pdev->dev, "Map iomem failed");
+        goto failed;
+    }
+    nifd->icap_base = nifd->nifd_base + 0x4000;
+
+    core = xocl_get_xdev(pdev);
+    if (!core)
+    {
+        xocl_err(&pdev->dev, "core is NULL in NIFD probe");
+        goto failed;
+    }
+    xocl_get_raw_header(core, &rom);
+    xocl_info(&pdev->dev, 
+            "NIFD: looking from NIFD in FeatureBitMap: %lx\n", 
+            (long)rom.FeatureBitMap);
+    nifd_valid = (long)rom.FeatureBitMap & 0x40000000;
+    if (!nifd_valid) {
+        return 0;
+    }
+
+    // cdev_init(&nifd->sys_cdev, &nifd_fops);
+    nifd->sys_cdev = cdev_alloc();
+    nifd->sys_cdev->ops = &nifd_fops;
+    nifd->sys_cdev->owner = THIS_MODULE;
+    nifd->instance = XOCL_DEV_ID(core->pdev) 
+                | platform_get_device_id(pdev)->driver_data;
+    nifd->sys_cdev->dev = MKDEV(MAJOR(nifd_dev), core->dev_minor);
+    err = cdev_add(nifd->sys_cdev, nifd->sys_cdev->dev, 1);
+    if (err)
+    {
+        xocl_err(&pdev->dev, "cdev_add failed, %d", err);
+        return err;
+    }
+
+    nifd->sys_device = device_create(xrt_class,
+                                     &pdev->dev,
+                                     nifd->sys_cdev->dev,
+                                     NULL,
+                                     "%s%d",
+                                     platform_get_device_id(pdev)->name,
+                                     nifd->instance);
+    if (IS_ERR(nifd->sys_device))
+    {
+        err = PTR_ERR(nifd->sys_device);
+        cdev_del(nifd->sys_cdev);
+        goto failed;
+    }
+
+    xocl_drvinst_set_filedev(nifd, nifd->sys_cdev);
+
+    platform_set_drvdata(pdev, nifd);
+    xocl_info(&pdev->dev, "NIFD device instance %d initialized\n",
+              nifd->instance);
+    return 0;
+
+failed:
+    xocl_drvinst_free(nifd);
+    return err;
+}
+
+static int nifd_remove(struct platform_device *pdev)
+{
+    struct xocl_nifd *nifd;
+    struct xocl_dev_core *core;
+
+    core = xocl_get_xdev(pdev);
+    if (!core)
+    {
+        xocl_err(&pdev->dev, "core is NULL in NIFD remove");
+    }
+
+    nifd = platform_get_drvdata(pdev);
+    if (!nifd)
+    {
+        xocl_err(&pdev->dev, "driver data is NULL");
+        return -EINVAL;
+    }
+    device_destroy(xrt_class, nifd->sys_cdev->dev);

```

Comment:
This might be crashing the system. Please Handle the probe early return case. One possible way is checking the resource availability before trying to destroy or free.

Reply:
@houlz0507 if the char device was not created, shouldn't it return at line 662? Is there an  API that I can use to check if create_device was called from pdev?

Reply:
No, it will not return on 662.  I think you may use nifd->sys_device != NULL to check if create_device() is success or not.

Reply:
@houlz0507 In the most recent commit, I changed it to only destroy the device if there is one. I will soon try that out on a few different systems. At the same time, please take a look if the change make sense.